src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -493,51 +493,105 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     }
 
     public static class Translog extends Timer implements ToXContent, Streamable {
-        private final AtomicInteger currentTranslogOperations = new AtomicInteger();
+        public static int UNKNOWN = -1;
 
-        public void reset() {
+        private int recovered;
+        private int total = UNKNOWN;
+        private int totalOnStart = UNKNOWN;
+
+        public synchronized void reset() {
             super.reset();
-            currentTranslogOperations.set(0);
+            recovered = 0;
+            total = UNKNOWN;
+            totalOnStart = UNKNOWN;
+        }
+
+        public synchronized void incrementRecoveredOperations() {
+            this.recovered++;
+            assert total == UNKNOWN || total >= recovered : "total, if known, should be > recovered. total [" + total + "], recovered [" + recovered + "]";
+        }
+
+        /** returns the total number of translog operations recovered so far */
+        public synchronized int recoveredOperations() {
+            return this.recovered;
+        }
+
+        /**
+         * returns the total number of translog operations needed to be recovered at this moment.
+         * Note that this can change as the number of operations grows during recovery.
+         * <p/>
+         * A value of -1 ({@link RecoveryState.Translog#UNKNOWN} is return if this is unknown (typically a gateway recovery)
+         */
+        public synchronized int totalOperations() {
+            return this.total;
+        }
+
+        public synchronized void totalOperations(int total) {
+            this.total = total;
+            assert total == UNKNOWN || total >= recovered : "total, if known, should be > recovered. total [" + total + "], recovered [" + recovered + "]";
         }
 
-        public void addTranslogOperations(int count) {
-            this.currentTranslogOperations.addAndGet(count);
+        /**
+         * returns the total number of translog operations to recovered, on the start of the recovery. Unlike {@link #totalOperations}
+         * this does change during recovery.
+         * <p/>
+         * A value of -1 ({@link RecoveryState.Translog#UNKNOWN} is return if this is unknown (typically a gateway recovery)
+         */
+        public synchronized int totalOperationsOnStart() {
+            return this.totalOnStart;
         }
 
-        public void incrementTranslogOperations() {
-            this.currentTranslogOperations.incrementAndGet();
+        public synchronized void totalOperationsOnStart(int total) {
+            this.totalOnStart = total;
         }
 
-        public int currentTranslogOperations() {
-            return this.currentTranslogOperations.get();
+        public synchronized float recoveredPercent() {
```

Comment:
can we maybe add a unittest for this percentage calculation? I also would love to use `getRecoveredPercent` instead of this pattern but I will open another issue for this to clean it up over the entire codebase